File: src/cpp_srvcli/src/add_three_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "tutorial_interfaces/srv/add_three_ints.hpp"

#include <chrono>
#include <memory>
#include <cstdlib>

using namespace std::chrono_literals;

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("add_three_ints_client");
  auto client = node->create_client<tutorial_interfaces::srv::AddThreeInts>("add_three_ints");

  auto request = std::make_shared<tutorial_interfaces::srv::AddThreeInts::Request>();
  request->a = atoll(argv[1]); // Convert the first argument to long long and assign it to request->a
  request->b = atoll(argv[2]); // Convert the second argument to long long and assign it to request->b
  request->c = atoll(argv[3]); // Convert the third argument to long long and assign it to request->c

  while (!client->wait_for_service(1s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(node->get_logger(), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(node->get_logger(), "Service not available, waiting again...");
  }

  auto future_result = client->async_send_request(request); // Send the request asynchronously

  // Wait for the result
  if (rclcpp::spin_until_future_complete(node, future_result) == rclcpp::FutureReturnCode::SUCCESS) {
    RCLCPP_INFO(node->get_logger(), "Result of add_three_ints: %ld", future_result.get()->sum);
  } else {
    RCLCPP_ERROR(node->get_logger(), "Failed to call service add_three_ints");
    return 1;
  }

  rclcpp::shutdown();
  return 0;
}

File: src/py_srvcli/py_srvcli/add_three_ints_client.py
import sys
from tutorial_interfaces.srv import AddThreeInts

import rclpy
from rclpy.node import Node

class AddThreeIntsClient(Node):

    def __init__(self):
        super().__init__('add_three_ints_client')
        self.cli = self.create_client(AddThreeInts, 'add_three_ints')
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Service not available, waiting again...')

    def send_request(self, a, b, c):
        req = AddThreeInts.Request()
        req.a = a
        req.b = b
        req.c = c
        self.future = self.cli.call_async(req)
        
        return self.future
    
def main():
    rclpy.init()

    add_three_ints_client = AddThreeIntsClient()
    future = add_three_ints_client.send_request(int(sys.argv[1]), int(sys.argv[2]), int(sys.argv[3]))
    rclpy.spin_until_future_complete(add_three_ints_client, future)

    if future.result() is not None:
        add_three_ints_client.get_logger().info(f'Result of add_three_ints: {future.result().sum}')
    else:
        add_three_ints_client.get_logger().error('Exception while calling service: %r' % future.exception())

    add_three_ints_client.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    if len(sys.argv) != 4:
        print("Usage: ros2 run py_srvcli add_three_ints_client <a> <b> <c>")
        sys.exit(1)
    main()
